src/NServiceBus.Core/Pipeline/Incoming/IncomingContext.cs

```diff
@@ -18,22 +17,26 @@ public abstract class IncomingContext : BehaviorContext, IMessageProcessingConte
         /// <summary>
         /// Initializes a new instance of <see cref="IncomingContext" />.
         /// </summary>
-        protected IncomingContext(BehaviorContext parentContext)
+        protected IncomingContext(string messageId, string replyToAddress, IReadOnlyDictionary<string, string> headers, BehaviorContext parentContext)
             : base(parentContext)
         {
+
+            this.MessageId = messageId;
+            this.ReplyToAddress = replyToAddress;
+            this.MessageHeaders = headers;
         }
 
         /// <inheritdoc />
         public ContextBag Extensions => this;
 
         /// <inheritdoc />
-        public string MessageId => Get<IncomingMessage>().MessageId;
+        public string MessageId { get; }
 
         /// <inheritdoc />
-        public string ReplyToAddress => Get<IncomingMessage>().GetReplyToAddress();
+        public string ReplyToAddress { get; }
```

Comment:
the `GetReplyToAddress` used by the physical context seems to be a TryGet. Replacing the property by a TryGet method just to indicate that the value is optional complicates stuff a lot (since we still need to provide it via constructor). I'd rather see a `Maybe<string>` property.

I tried to keep the impact of that PR as small as possible, since currently this is only done to improve testing. You can call  `GetReplyToAddress` (not TryGetReplyToAddress) anywhere too.